Building an API pt. 8 - lesson 2: 'One-to-Many: Nested Routes' (#171)
Ada-Developers-Academy/core-unit-2: api-8-relationships-in-sqlalchemy/nested-routes-in-flask.md

  - `POST` to `/authors`

> The `Book` model and table should have the following columns:
> - `id`
> - `title`
> - `description`
> - `author_id`
> - `author` (model only)
> 
> The `Author` model and table should have the following columns:
> - `id`
> - `name`
> - `books` (model only)
> 
> `route_utilities.py` should contain:
> - The function `validate_model()` that can retrieve a model of any type
> 

Reviewer: Formatting wise, this expansion feels inconsistent with indentation and bullet points, specifically on lines 37, 44 and 49. My suggestion would be: 

```suggestion
- The following columns in the `Book` model and table :
     - `id`
     - `title`
     - `description`
     - `author_id`
     - `author` (model only)

-  The following columns in the `Author` model and table:
     - `id`
     - `name`
     - `books` (model only)

- The following function in `route_utilities.py`:
     - The function `validate_model()` that can retrieve a model of any type
```

Author: Fully agree that the existing formatting is not consistent across the features. What are your feelings on the lesson-to-lesson consistency tradeoff if the feature drop downs only have this formatting starting from this lesson forward (in case there isn't time to update the drop downs in the prior Building an API lessons)? I'm cool making the change as long as the shift at a certain point in the series doesn't bug anyone.

## More Than One Right Answer

There's more than one way to implement the relationship between our `Book` and `Author` models in our routes. For our curriculum, we will stick with RESTful routing naming conventions and implement a *nested route* to describe this relationship.

### !end-callout

### Nested Routes

Nested routes (also called nested resources) describe the parent-child relationship between (at least) two models in our API. By nesting resources in a single route, the client can retrieve only the data they require. For example, we can read all `book`s (`GET`) or create (`POST`) a `book` by a specific `author`.

![A diagram of a nested route: '/author/author_id/books'](../assets/nested-routes-in-flask_route-diagram.png)  
*Fig.  A diagram of a nested route*

The nested route starts with the _parent_ model. In this case, `authors`. Then we specify the appropriate author record's `id` we want our new _child_ model record to belong to, indicated by the `<author_id>` parameter. Finally, we list the name of the child collection, in this case, `books`.

### Creating Our Endpoint

> Our nested route `POST` `/authors/<author_id>/books` will use the `bp` blueprint in `author_routes.py`. We want our `POST` endpoint to return a new `Book` record, so we should also consider adjusting the convenience functions we created for the `Book` model, `to_dict` and `from_dict`, to handle an optional `Author`. 

Reviewer: "Call out adjusting the Book convenience functions to handle optional Author" ✅

- `CREATE`ing a book: We could choose to change our `create_book` route to account for `Author` information, but there is no need to since we made a design decision for this project to create a nested route to represent creating a book with an author.

- `GET`ting books: By updating our `Book` convenience methods, our routes `get_all_books` and `get_one_book` already account for `Author` information, which is great! 

- `UPDATE`ing a book: Our `update_book` route will only set the `title` and `description` of a book currently. What about if the author information is missing or incorrect? We have a great opportunity to update this route and improve the functionality of our API! 

- `DELETE`ing a book: Our `delete_book` route will work to remove the book record, whether or not is has `Author` information attached.

We have not implemented changes to the `update_book()` route to account for `Author` information in the `hello-books` repo, but have left it as an exercise for folks to try out independently. See what you come up with and feel free to share and discuss your improvements in Slack!

> 

Reviewer: "Can leave as an exercise for students to ask them how they would change the update_book  route so we can change the author if it's missing or incorrect." ✅

## Check for Understanding

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: b4eb362c-cfb1-41b1-9fbb-f83d2cd3c228
* title: One-to-Many: Nested Routes
##### !question

Select all options below which are true about nested routes.

##### !end-question
##### !options

a| They describe a parent-child relationship between two or more models.
b| A nested route starts with the child collection.
c| The parent resource's id is not identified in a nested route.
d| By nesting resources in a single route, the client can retrieve only the data they require.
e| A nested route ends with the child collection.

##### !end-options
##### !answer

a|
d|
e|

##### !end-answer
##### !explanation

* True.
* False, A nested route starts with the parent collection.
* False, The parent resource's id is typically the second item in a nested route.
* True.
* True.

> ##### !end-explanation

Reviewer: "Replace CFU question" ✅

I like this change to the Check for Understanding!
